### DISCUSSION TITLE
How to set the hardening factor as a function of variables in J2 plasticity model

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#19454] [Q&A General] [answered]
Dear MOOSE experts,

I tried to replace the linear elastic mechanics in [bicrysta.i](https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/combined/test/tests/ACGrGrElasticDrivingForce/bicrystal.i) with J2 plastic mechanics, but I encountered a very big bottleneck, that is, I don’t know how to set the `HardFactor` in `userobjects` to the phase field variable `gr0 gr1` stand up. And try to set `HardFactor` to `MaterialProperty<Real>`, but the compilation error (the transmitted data type is inconsistent with the output data type), the error is displayed as follows,

![6C~%{ TBO(QA(OLM%X$HZK2](https://user-images.githubusercontent.com/56183775/142956183-5354b29e-7808-428d-8cdd-baf984371d74.png)

And the following is the linear hardening model that I customized from [TensorMechanicsHardeningPowerRule.C](https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/tensor_mechanics/src/userobjects/TensorMechanicsHardeningPowerRule.C),
```c++
\\\ TensorMechanicsHardeningLinear.h
public:
  static InputParameters validParams();

  TensorMechanicsHardeningLinear(const InputParameters & parameters);

  virtual Real value(Real intnl) const override;

  virtual Real derivative(Real intnl) const override;

private:
  /// The value = value_0 + _h_factor*p
  const Real _value_0;

  /// The value = value_0 + _h_factor*p
  const Real _hard_factor;

\\\ TensorMechanicsHardeningLinear.C
Real
TensorMechanicsHardeningLinear::value(Real intnl) const
{
  return _value_0 + _hard_factor * intnl;
  // return _value_0 * std::pow(intnl / _epsilon0 + 1, _exponent);
}

Real
TensorMechanicsHardeningLinear::derivative(Real intnl) const  // d*/d_intnl
{
  return _hard_factor;
}
```

In general, my purpose is to set HardFactor to a function of `gr0 gr1`, just like `ElasticityTensor`, so what I want to do now is to set HardFactor to `MaterialProperty<Real>`

Any suggestions or recommendations to fix the problem would be greatly appreciated.

Thank you
Wei

**Comment by GiudGiud** [comment, marked as answer]
`_qp` is defined for regular materials using the computeQpProperties routine. This is being executed every time we `enter` a new element and need the material properties for a kernel.

here this ` TensorMechanicsHardeningPowerRule` is a user object, and more precisely it derives from a GeneralUO. GeneralUO isnt meant to be executed everywhere on the mesh, or at quadrature points (_qp) but only once for the entire domain. 

If you go in framework/include/postprocessors and enter `grep -r GeneralPostprocessor` (which derives from GeneralUO) you can see the general style of a GeneralUO derived class.

So my suggestion would be to look again at what you want value to return. Is it a single value for all the mesh or is it a field? If the latter you need to change the inheritances.

**Reply by PengWei97** [reply]
> So my suggestion would be to look again at what you want value to return. Is it a single value for all the mesh or is it a field? If the latter you need to change the inheritances.

Yes, I hope the returned value is a field, specifically for different `gr0 gr1`, HardFactor has different values. HardFactor is a `MaterialProperty<Real>` data, so I need to change my inheritance from `GeneralUserObject` to `Material`, right?

**Reply by GiudGiud** [reply]
if you want to declare a material property for _hard_factor yes, that's what I would do.

**Reply by PengWei97** [reply]
But there is a big problem here. If I put `TensorMechanicsHardeningLinear` in the materials module of input, there is a big problem you will encounter, that is, `TensorMechanicsPlasticJ2` is in `UserObjects`, and the part of `ComputeMultiPlasticityStress`  in `materials`  about calling the yield function needs to be modified. I think this is a very big change. I am not sure if such a big change must be made to migrate the the J2 plastic model from `UserObjects` to `materials` block.
The following is the relevant code involved in the input file,
```bash
[Materials]
  [./mc]
    type = ComputeMultiPlasticityStress
    block = 0
    ep_plastic_tolerance = 1E-9
    plastic_models = j2
    debug_fspb = crash
  [../]
[./]

[UserObjects]
  [./str]
    type = TensorMechanicsHardeningLinear
    value_0 = 2000 # MPa
    HardFactor = ${my_HardFactor}  # MPa
  [../]
  [./j2]
    type = TensorMechanicsPlasticJ2
    yield_strength = str
    yield_function_tolerance = 1E-5
    internal_constraint_tolerance = 1E-9
    max_iterations = 10
  [../]
```

**Reply by GiudGiud** [reply]
ok someone more familiar with this system will have to pitch in then

**Reply by PengWei97** [reply]
Yes, because of the modification of `TensorMechanicsHardeningLinear`, the related code of `TensorMechanicsPlasticJ`2` and `ComputeMultiPlasticityStress` had to be modified afterwards. This is a relatively large project.

Specifically, I am not very clear how to modify the following code to call `MaterialProperty`, if it is in the `materials` block I could do.
```c++
\\\TensorMechanicsPlasticJ2.C
_strength(getUserObject<TensorMechanicsHardeningModel>("yield_strength")),
```
And, I think the biggest challenge is in `ComputeMultiPlasticityStress`, because it involves three levels of inheritance to see the call of the plastic model (as shown in the figure below), and how to replace this plastic_models with MaterialProperty data.
![image](https://user-images.githubusercontent.com/56183775/142988041-d0e754bc-04d7-4899-a58e-9df0ba170054.png)

Of course, this is just my understanding to the best of my knowledge. I would be very grateful if relevant experts can give me guidance in this regard.
